request-manager/graph/grapher.go

```diff
@@ -31,13 +34,16 @@ func NewGrapher(specs spec.Specs, idGenFactory id.GeneratorFactory) *Grapher {
 // DoChecks performs graph checks for all sequences.
 func (gr *Grapher) DoChecks() (seqGraphs map[string]*Graph, seqErrors map[string]error) {
 	seqGraphs = map[string]*Graph{}
-	seqSets := map[string]map[string]bool{}
 	seqErrors = map[string]error{}
 
+	// sequence name -> set of job args that its nodes declare in `sets` field
+	// i.e. the job args that the sequence sets
+	seqSets := map[string]map[string]bool{}
+
 	// First, build all the sequence graphs.
 	// Since sequence graphs are just a graph representation of dependencies
-	// between a single sequence's nodes, this is per sequence; there's no
-	// need to look at any subsequences.
+	// between a single sequence's nodes, this is contained to one sequence;
+	// this won't look at any subsequences.
 	for seqName, seqSpec := range gr.allSequences {
 		// Generates IDs unique within sequence graph
 		idgen := gr.idGenFactory.Make()
```

Comment:
I think at this point we do not need to generate IDs? Sequence graphs are (to use the previous term) templates (or prototypes, if you like), so they're generic. It's not until we resolve them into request graphs that they (and job nodes) need unique IDs (unique only within a request graph, not universally unique).

Reply:
Sequence graph nodes need to be keyed on something. The alternative here is to just use the node name, which is unique within the sequence graph. I left it up to a UID generator because it seems better for the perhaps unlikely case where the sequence graph is extremely large, and the node names are inconvenient for whatever golang's default string hash function is (I can't seem to find the answer in the first three links of my first google search). For small graphs, maybe the overhead is greater than if I just keyed on node names, but it's not noticeably slow for our requests dir, so I left it.

Reply:
Ok, this is fine for now. My concern was that it's a bit confusing because at this point (i.e. in a SG), the concept of a node ID _slightly_ conflicts with the real node IDs which are part of request graphs. In an SG, nodes don't need a unique ID because they already have 2 unique identifies: name and spec order (not to be confused with resulting graph topological order). So in addition to slightly conflict with real node IDs, they're also not strictly needed here.

But we can leave it for now because this is better addressed later if/when we make a new `Graph` that does a better job of hiding all such graph details.